streamingccl: add generation support for sinkless stream

annezhu98 opened the pull request:
See [rfc](https://github.com/cockroachdb/cockroach/pull/56932/files) for more context.

Part of [Epic CRDB-6407](https://cockroachlabs.atlassian.net/browse/CRDB-6407)

Informs: #57406

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67189)
<!-- Reviewable:end -->

azhu-crl committed:
streamingccl: replan dist flow on generation events

Once the coordinator node receives a `GenerationEvent` from its client, dist flow will be replanned with the topology associated with the new generation.

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=67189) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=67189) before we can accept your contribution.<br/><hr/>**Anne Zhu** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=67189) it.</sub>
